### PR TITLE
[clojure][generator][WIP] add random namespace generation

### PR DESCRIPTION
**Pull request by Chouffe**
PR moved here: https://github.com/apache/incubator-mxnet/pull/14800

## Description ##

Split random functions in their own generated namespace.

* `ndarray_random_api`
* `symbol_random_api`

It renames the following functions to be as close as possible to the scala names:
* `-random-randint` into `randint` 
* `sample-randint` into `randint-like`for instance and put them under `ndarray-random-api` namespace.

## TODO

* Unify random and sample functions (mu vs loc and sigma)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by Chouffe** [comment]
cc @kedarbellare :-)

File: contrib/clojure-package/src/dev/generator.clj [hunk 1]
```diff
@@ -470,6 +476,53 @@
   (println "Generating symbol-api file")
   (write-to-file all-symbol-api-functions symbol-api-gen-ns "src/org/apache/clojure_mxnet/gen/symbol_api.clj"))
 
+;;;;;;; SymbolRandomAPI
+
+(defn gen-symbol-random-api-function [op-name]
```
Comment (kedarbellare) [inline contrib/clojure-package/src/dev/generator.clj]: is it possible to share code between symbol api and symbol random api generation?

File: contrib/clojure-package/src/dev/generator.clj [hunk 2]
```diff
@@ -470,6 +476,53 @@
...
+  (let [{:keys [fn-name fn-description args]} (gen-op-info op-name)
+        fn-name (remove-prefix "-random-" fn-name)
```
Comment (kedarbellare) [inline contrib/clojure-package/src/dev/generator.clj]: the scala package does it for both `random_` and `sample_` (see: https://github.com/apache/incubator-mxnet/blob/master/scala-package/macros/src/main/scala/org/apache/mxnet/GeneratorBase.scala#L180).

File: contrib/clojure-package/src/dev/generator.clj [hunk 3]
```diff
@@ -470,6 +476,53 @@
...
+
+(def symbol-random-api-gen-ns "(ns
+  ^{:doc \"Experimental\"}
+  org.apache.clojure-mxnet.symbol-random-api
+  (:refer-clojure :exclude [* - + > >= < <= / cast concat identity flatten load max
+                            min repeat reverse set sort take to-array empty sin
+                            get apply shuffle ref])
+  (:require [org.apache.clojure-mxnet.util :as util]
+            [org.apache.clojure-mxnet.shape :as mx-shape])
+  (:import (org.apache.mxnet SymbolAPI)))")
```
Comment (kedarbellare) [inline contrib/clojure-package/src/dev/generator.clj]: we should use the `SymbolRandomAPI` instead (see: http://mxnet.incubator.apache.org/api/scala/docs/index.html#org.apache.mxnet.SymbolRandomAPI$)

File: contrib/clojure-package/src/dev/generator.clj (hunk 2 above)
Comment (Chouffe) [reply contrib/clojure-package/src/dev/generator.clj]: Do we also want a `ndarray-sample` namespace or should we use a prefix like `sample-op-name` in the `ndarray-random` namespace?

File: contrib/clojure-package/src/dev/generator.clj (hunk 1 above)
Comment (Chouffe) [reply contrib/clojure-package/src/dev/generator.clj]: Yes it is :) I am refactoring this now

**Comment by Chouffe** [comment]
@kedarbellare 
1. Done
2. I could not get this to work when calling directly the `NDArrayRandomAPI` and the `SymbolRandomAPI`
3. This is not done yet. I am not entirely sure how to do that. Can it be done in a subsequent PR?
4. Done

**Comment by Chouffe** [comment]
cc @gigasquid for feedback on the `generator` namespace. I am not yet comfortable with the code there ^^

File: contrib/clojure-package/src/dev/generator.clj (hunk 3 above)
Comment (Chouffe) [reply contrib/clojure-package/src/dev/generator.clj]: I could not get it to work. Do you have any pointers on how this should be done?

**Comment by Chouffe** [comment]
I reopened a PR here: https://github.com/apache/incubator-mxnet/pull/14800

**Comment by roywei** [comment]
@mxnet-label-bot add [Clojure, pr-work-in-progress]

**Pull request closed without merging**